File: api-scripts/services/cloudflare.py
import requests
from config import CLOUDFLARE_API_URL, CLOUDFLARE_API_KEY, CLOUDFLARE_EMAIL


class CloudflareClient:
    """Cloudflare DNS API client using Global API Key auth."""

    def __init__(self, api_key: str = None, email: str = None):
        self.api_key = api_key or CLOUDFLARE_API_KEY
        self.email = email or CLOUDFLARE_EMAIL
        self._zone_cache = {}

    def _headers(self) -> dict:
        return {
            "X-Auth-Key": self.api_key,
            "X-Auth-Email": self.email,
            "Content-Type": "application/json",
        }

    def _request(self, method: str, path: str, **kwargs) -> dict:
        url = f"{CLOUDFLARE_API_URL}{path}"
        resp = requests.request(method, url, headers=self._headers(), **kwargs)
        data = resp.json()
        if not data.get("success", False):
            errors = data.get("errors", [])
            raise RuntimeError(f"Cloudflare {method} {path} failed: {errors}")
        return data

    def get_zone_id(self, domain: str) -> str:
        """Get zone ID for a domain."""
        if domain in self._zone_cache:
            return self._zone_cache[domain]
        data = self._request("GET", f"/zones?name={domain}")
        results = data.get("result", [])
        if not results:
            raise RuntimeError(f"Zone not found for domain: {domain}")
        zone_id = results[0]["id"]
        self._zone_cache[domain] = zone_id
        return zone_id

    def list_dns_records(self, domain: str, type: str = None, name: str = None) -> list:
        """List DNS records for a zone."""
        zone_id = self.get_zone_id(domain)
        params = {}
        if type:
            params["type"] = type
        if name:
            params["name"] = name
        data = self._request("GET", f"/zones/{zone_id}/dns_records", params=params)
        return data.get("result", [])

    def create_dns_record(self, domain: str, record_type: str, name: str,
                          content: str, ttl: int = 3600, priority: int = None,
                          proxied: bool = False) -> dict:
        """Create a DNS record."""
        zone_id = self.get_zone_id(domain)
        payload = {
            "type": record_type,
            "name": name,
            "content": content,
            "ttl": ttl,
            "proxied": proxied,
        }
        if priority is not None:
            payload["priority"] = priority
        data = self._request("POST", f"/zones/{zone_id}/dns_records", json=payload)
        return data.get("result", {})

    def upsert_dns_record(self, domain: str, record_type: str, name: str,
                          content: str, ttl: int = 3600, priority: int = None,
                          proxied: bool = False) -> dict:
        """Create or update a DNS record. If a record of the same type+name
        exists, update it; otherwise create it."""
        existing = self.list_dns_records(domain, type=record_type, name=name)
        zone_id = self.get_zone_id(domain)
        payload = {
            "type": record_type,
            "name": name,
            "content": content,
            "ttl": ttl,
            "proxied": proxied,
        }
        if priority is not None:
            payload["priority"] = priority

        if existing:
            record_id = existing[0]["id"]
            data = self._request("PUT", f"/zones/{zone_id}/dns_records/{record_id}",
                                 json=payload)
        else:
            data = self._request("POST", f"/zones/{zone_id}/dns_records", json=payload)
        return data.get("result", {})

File: microsoft works/src/cloudflare.js
const log = require("./logger");
const { getAccessToken } = require("./mfa");

/**
 * Add a DNS record via Cloudflare API.
 */
async function addDnsRecord(apiKey, zoneId, type, name, content, priority, ttl = 3600) {
  const url = `https://api.cloudflare.com/client/v4/zones/${zoneId}/dns_records`;
  const body = { type, name, content, ttl };
  if (priority !== undefined && priority !== null) body.priority = priority;
  if (type === "CNAME") body.proxied = false;

  const res = await fetch(url, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${apiKey}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify(body),
  });

  const data = await res.json();
  if (data.success) {
    log.success(`  DNS ${type} record added: ${name} -> ${content}`);
    return data.result;
  }

  // Check if record already exists
  const errors = data.errors || [];
  const alreadyExists = errors.some(
    (e) => e.code === 81057 || (e.message && e.message.includes("already exists"))
  );
  if (alreadyExists) {
    log.warn(`  DNS ${type} record already exists: ${name}`);
    return null;
  }

  throw new Error(`Cloudflare DNS error: ${JSON.stringify(errors)}`);
}

/**
 * Get the tenant's .onmicrosoft.com domain from Graph API.
 */
async function getTenantOnmicrosoftDomain(tenantId, clientId, clientSecret) {
  const token = await getAccessToken(tenantId, clientId, clientSecret);
  const res = await fetch("https://graph.microsoft.com/v1.0/domains", {
    headers: { Authorization: `Bearer ${token}` },
  });

  if (!res.ok) {
    throw new Error(`Failed to get domains: ${res.status}`);
  }

  const data = await res.json();
  const domains = data.value || [];
  const onmicrosoft = domains.find(
    (d) => d.id && d.id.endsWith(".onmicrosoft.com") && !d.id.includes(".mail.")
  );

  if (!onmicrosoft) {
    throw new Error("Could not find .onmicrosoft.com domain");
  }

  // Return just the subdomain part (e.g. "contoso" from "contoso.onmicrosoft.com")
  return onmicrosoft.id.replace(".onmicrosoft.com", "");
}

/**
 * Add all M365 DNS records to Cloudflare for a domain.
 * Records: MX, SPF, Autodiscover CNAME, DKIM CNAME x2, DMARC TXT
 */
async function addAllM365Records(domain, apiKey, zoneId, tenantOnmicrosoft) {
  log.info(`Adding M365 DNS records for ${domain}...`);

  // Convert domain dots to dashes for DKIM (e.g. "example.com" -> "example-com")
  const domainDashes = domain.replace(/\./g, "-");

  const records = [
    // MX record
    {
      type: "MX",
      name: domain,
      content: `${domain}.mail.protection.outlook.com`,
      priority: 0,
    },
    // SPF TXT record
    {
      type: "TXT",
      name: domain,
      content: "v=spf1 include:spf.protection.outlook.com -all",
      priority: null,
    },
    // Autodiscover CNAME
    {
      type: "CNAME",
      name: `autodiscover.${domain}`,
      content: "autodiscover.outlook.com",
      priority: null,
    },
    // DKIM CNAME selector1
    {
      type: "CNAME",
      name: `selector1._domainkey.${domain}`,
      content: `selector1-${domainDashes}._domainkey.${tenantOnmicrosoft}.onmicrosoft.com`,
      priority: null,
    },
    // DKIM CNAME selector2
    {
      type: "CNAME",
      name: `selector2._domainkey.${domain}`,
      content: `selector2-${domainDashes}._domainkey.${tenantOnmicrosoft}.onmicrosoft.com`,
      priority: null,
    },
    // DMARC TXT record
    {
      type: "TXT",
      name: `_dmarc.${domain}`,
      content: `v=DMARC1; p=quarantine; rua=mailto:admin@${domain}`,
      priority: null,
    },
  ];

  let successCount = 0;
  for (const rec of records) {
    try {
      await addDnsRecord(apiKey, zoneId, rec.type, rec.name, rec.content, rec.priority);
      successCount++;
    } catch (err) {
      log.error(`  Failed to add ${rec.type} ${rec.name}: ${err.message}`);
    }
  }

  log.info(`  ${successCount}/${records.length} DNS records processed for ${domain}`);
  return successCount;
}

module.exports = {
  addDnsRecord,
  getTenantOnmicrosoftDomain,
  addAllM365Records,
};
